### PR TITLE
Introduced restrictive logging of sensitive data and configuration of services.

### PR DESCRIPTION
**Pull request by Shreyas-vgr**
- feat: Introduced restrictive logging configuration through environment variables to mask sensitive data information from being logged.
- feat: Introduced new prop on ControlHandler and ControlManager : `services` to customize and override default service implementations on the framework
- `services` can be extended to include and customize i18 resources etc

## Description
- This PR introduces a new interface `ControlServices` which provides static methods `ControlServices.setDefaults()` where the developer can use to set their custom services implementation such as override our default logging implementation.
`services` are added as an optional parameter to `ControlHandler` and `ControlManager` constructors if specific service overrides are requires bases on use cases.
- We have also introduced `restrictiveLogging` to mask sensitive information from being logged which can be configured by setting an environment variable `"ASK_SDK_RESTRICTIVE_LOGGING": "true"`.

## Testing
- Tested our log messages are still consistent with changes by running npm scripts : `test, format-check etc`
- Created a sample override implementation in `NumberControl demo` to demonstrate the usage of ControlServices
- All existing test cases passes.

## Screenshots (if appropriate)
- Local debugging logging information by default
Screen Shot 2021-05-20 at 3 20 16 PM

- Local debugging of a demo skill
 with Restrictive Logging set on environment variables
Screen Shot 2021-05-20 at 3 19 25 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Docs(Add new document content)
- [ ] Translate Docs(Translate document content)

## Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] My commit message follows [Conventional Commit Guideline](https://conventionalcommits.org/)

## License
- [x] By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

[issues]: https://https://github.com/alexa/ask-sdk-controls/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: src/controls/interfaces/ILogger.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    info(message: string): void;
+
+    warn(message: string): void;
+
+    debug(message: string): void;
+
+    logObject(level: string, id: string, object: any, stringify?: boolean): void;
+
+    sanitize(message: string): string;
```
Comment (mikelid) [inline src/controls/interfaces/ILogger.ts:31]: Have all log messages that are risky been wrapped with a call to sanitize? e.g.
`this.log.info(`Slot value = ${this.log.sanitize(slotValue)}`

File: src/logging/DefaultLogger.ts [hunk 2]
```diff
@@ -75,4 +76,42 @@ export class Logger {
...
+    /**
+     * Log an object with pretty print formatting and also
+     * masking sensitive information if required
+     *
+     * @param id - Unique label.
+     * @param object - Object to be logged
+     */
+    logObject(logLevel: string, id: string, object: any, stringify?: true): void {
+        const objectToLogJson: string = stringify !== true ? object : JSON.stringify(object, null, 2);
+        const message = this.sanitize(objectToLogJson);
```
Comment (mikelid) [inline src/logging/DefaultLogger.ts:89]: I don't think we want to sanitize all objects by default here.  the issue is that a different implementation of logObject may not do it and then we are hosed.  Suggest hoist this to all the call sites of logObject... then the two problems (object logging vs sanitize) are kept orthogonal.

**Review by mikelid — CHANGES_REQUESTED** [review]

File: demo/ComponentModeAPL/src/buildInteractionModel.ts [hunk 3]
```diff
@@ -12,10 +12,12 @@
  */
 
 import { ComponentModeDemo } from '.';
+import { ControlServices } from '../../../src/controls/ControlServices';
 import { ControlInteractionModelGenerator } from '../../../src/interactionModelGeneration/ControlInteractionModelGenerator';
-import { Logger } from '../../../src/logging/Logger';
 
-const log = new Logger('ComponentModeDemo:InteractionModel');
+const MODULE_NAME = 'ComponentModeDemo:InteractionModel';
+const services = ControlServices.getDefaults();
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply demo/ComponentModeAPL/src/buildInteractionModel.ts]: Yep, exported a helper method `ControlServices.getLogger(namespace)`

File: demo/NumberControl/NumberDemo/src/index.ts [hunk 4]
```diff
@@ -1,10 +1,38 @@
...
 
+export class CustomLoggerFactory implements ILoggerFactory {
+    getLogger(namespace: string): ILogger {
+        return new CustomLogger(namespace);
+    }
+}
+
+export class CustomLogger extends DefaultLogger {
+    info(message: string): void {
+        Debug(`info:Redfox:${this.moduleName}`)(message);
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply demo/NumberControl/NumberDemo/src/index.ts]: Updated

File: src/controls/ControlManager.ts [hunk 5]
```diff
@@ -121,15 +121,29 @@ export abstract class ControlManager implements IControlManager {
...
+     * The logger implementation.
+     */
+    log: ILogger;
+
     /**
      * Creates an instance of a Control Manager.
      * @param props - props
      */
-    constructor(props?: ControlManagerProps) {
+    constructor(props?: ControlManagerProps, services?: ControlServicesProps) {
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply src/controls/ControlManager.ts]: Moved to ControlManagerProps

File: src/controls/interfaces/ILogger.ts (hunk 1 above)
Comment (Shreyas-vgr) [reply src/controls/interfaces/ILogger.ts]: Yes, all the log messages containing sensitive information are being wrapped with sanitize methods

File: src/logging/DefaultLogger.ts (hunk 2 above)
Comment (Shreyas-vgr) [reply src/logging/DefaultLogger.ts]: You're right, hoisted `this.sanitize()` to all call sites of logObject

File: src/controls/interfaces/ILogger.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+/**
+ * Defines the Logger interface to define the logging format and
+ * log level used across the framework.
+ */
+export interface ILogger {
+    moduleName: string;
+
+    error(message: string): void;
```
Comment (sattpat) [inline src/controls/interfaces/ILogger.ts:21]: Nit: can you add some doc strings for the interface?

Comment (Shreyas-vgr) [reply src/controls/interfaces/ILogger.ts:25]: Updated!

**Review by sattpat — APPROVED** [review]
:shipit:

**Pull request merged**